File: rules/nostr/nip-44/missing-version-check.ts
// Fixtures for missing-version-check-ts
// Run: semgrep --test rules/nostr/nip-44/

import { base64 } from "@scure/base";

// BAD: slices past the version byte without checking it.
function nip44DecryptBad(payload: string): Uint8Array {
  const bytes = base64.decode(payload);
  // ruleid: missing-version-check-ts
  const nonce = bytes.slice(1, 33);
  const ciphertext = bytes.slice(33, bytes.length - 32);
  void nonce;
  return ciphertext;
}

// OK: explicit version byte check before slicing.
function nip44DecryptGood(payload: string): Uint8Array {
  const bytes = base64.decode(payload);
  if (bytes[0] !== 0x02) {
    throw new Error(`unsupported NIP-44 version: ${bytes[0]}`);
  }
  // ok: missing-version-check-ts
  const nonce = bytes.slice(1, 33);
  return nonce;
}

// BAD: function whose name matches decrypt|decode|parse|unwrap and slices past
// byte 0 without checking — even with a generic name.
function decodePayload(payload: string): Uint8Array {
  const bytes = base64.decode(payload);
  // ruleid: missing-version-check-ts
  const body = bytes.slice(1);
  return body;
}

// OK: alias-then-check pattern (the idiomatic style in nostr-tools).
function nip44DecodeAliased(payload: string): Uint8Array {
  const bytes = base64.decode(payload);
  const vers = bytes[0];
  if (vers !== 2) {
    throw new Error("unknown encryption version " + vers);
  }
  // ok: missing-version-check-ts
  const nonce = bytes.slice(1, 33);
  return nonce;
}

// OK: function name does NOT match decrypt|decode|parse|unwrap, so rule does not apply
// (Example: ECDH shared-secret extraction in getConversationKey style.)
function getConversationKey(secret: Uint8Array): Uint8Array {
  const sharedPoint = secret;
  // ok: missing-version-check-ts
  const xCoord = sharedPoint.subarray(1, 33);
  return xCoord;
}

File: rules/nostr/nip-44/missing-version-check.rs
// Fixtures for nostr.nip-44.missing-version-check-rust
// Run: semgrep --test rules/nostr/nip-44/

use base64::{engine::general_purpose::STANDARD, Engine};

/// BAD: slices past the version byte without checking it.
fn nip44_decrypt_bad(payload: &str) -> Result<Vec<u8>, anyhow::Error> {
    let bytes = STANDARD.decode(payload)?;
    // ruleid: nostr.nip-44.missing-version-check-rust
    let nonce = &bytes[1..33];
    let ciphertext = &bytes[33..bytes.len() - 32];
    let _mac = &bytes[bytes.len() - 32..];
    let _ = nonce;
    Ok(ciphertext.to_vec())
}

/// OK: explicit version byte check before slicing.
fn nip44_decrypt_good(payload: &str) -> Result<Vec<u8>, anyhow::Error> {
    let bytes = STANDARD.decode(payload)?;
    if bytes[0] != 0x02 {
        return Err(anyhow::anyhow!("unsupported NIP-44 version: {}", bytes[0]));
    }
    // ok: nostr.nip-44.missing-version-check-rust
    let nonce = &bytes[1..33];
    Ok(nonce.to_vec())
}

/// OK: validation through `.first()`.
fn nip44_decrypt_first(payload: &str) -> Result<Vec<u8>, anyhow::Error> {
    let bytes = STANDARD.decode(payload)?;
    if bytes.first() != Some(&0x02) {
        return Err(anyhow::anyhow!("unsupported NIP-44 version"));
    }
    // ok: nostr.nip-44.missing-version-check-rust
    let nonce = &bytes[1..33];
    Ok(nonce.to_vec())
}

/// BAD: any function inside a nip44-path file whose name matches the
/// decrypt|decode|parse|unwrap regex and slices past byte 0 without checking it.
fn decode_payload(payload: &str) -> Result<Vec<u8>, anyhow::Error> {
    let bytes = STANDARD.decode(payload)?;
    // ruleid: nostr.nip-44.missing-version-check-rust
    let body = &bytes[1..];
    Ok(body.to_vec())
}

/// OK: alias-then-check pattern (idiomatic in real codebases).
fn nip44_decrypt_aliased(payload: &str) -> Result<Vec<u8>, anyhow::Error> {
    let bytes = STANDARD.decode(payload)?;
    let version = bytes[0];
    if version != 0x02 {
        return Err(anyhow::anyhow!("unsupported NIP-44 version: {}", version));
    }
    // ok: nostr.nip-44.missing-version-check-rust
    let nonce = &bytes[1..33];
    Ok(nonce.to_vec())
}

/// OK: function name does NOT match decrypt|decode|parse|unwrap, so rule does not apply
/// even though it slices past byte 0. (Example: ECDH shared-secret extraction.)
fn get_conversation_key(secret_key: &[u8], pubkey: &[u8]) -> Vec<u8> {
    let shared = STANDARD.decode("foo").unwrap_or_default();
    // ok: nostr.nip-44.missing-version-check-rust
    let x_coord = &shared[1..33];
    x_coord.to_vec()
}
